File: Java/binnarytree/zigzagtraversal.java
package binnarytree;

import java.util.Map;
import java.util.Scanner;
import java.util.Stack;

public class zigzagtraversal {

	public class BinaryTree {

		private class Node {
			int data;
			Node left;
			Node right;

			Node(int data, Node left, Node right) {
				this.data = data;
				this.left = left;
				this.right = right;

			}

		}

		private Node root = null;
		private int size = 0;

		BinaryTree() {
			Scanner scn = new Scanner(System.in);
			this.root = input(scn, null, false);

		}

		private Node input(Scanner scn, Node parent, Boolean isleftorright) {
			if (parent == null) {
				System.out.println("Enter data for the parent node");
			} else {

				if (isleftorright) {
					System.out.println("Enter data for the left child of " + parent.data);
				} else {
					System.out.println("Enter data for the right child of " + parent.data);
				}
			}

			int data = scn.nextInt();
			Node node = new Node(data, null, null);
			this.size++;
			System.out.println("does " + node.data + " has left child");
			Boolean choice = false;
			choice = scn.nextBoolean();
			if (choice) {
				node.left = input(scn, node, true);
			}

			choice = false;
			System.out.println("does " + node.data + " has right child");
			choice = scn.nextBoolean();
			if (choice) {
				node.right = input(scn, node, false);
			}

			return node;

		}

		public void display() {
			display(this.root);
		}

		private void display(Node parent) {

			String str = "";
			if (parent.left != null) {
				str += parent.left.data + " => ";
			} else {
				str += "End => ";
			}
			str += parent.data;
			if (parent.right != null) {
				str += " <= " + parent.right.data;
			} else {
				str += " <= End";
			}
			System.out.println(str);
			if (parent.left != null) {
				display(parent.left);
			}
			if (parent.right != null) {
				display(parent.right);
			}

		}

		public void zigzag() {
			zigzag(this.root);
		}

		private void zigzag(Node parent) {
			Stack<Node> p = new Stack<>();
			Stack<Node> c = new Stack<>();
			p.push(parent);
			int level = 0;
			while (!p.isEmpty()) {
				Node node = p.pop();
				System.out.println(node.data + " ");
				if (level % 2 == 0) {
					if (node.left != null) {
						c.push(node.left);
					}
					if (node.right != null) {
						c.push(node.right);
					}
				} else {
					if (node.right != null) {
						c.push(node.right);
					}
					if (node.left != null) {
						c.push(node.left);
					}
				}
				if (p.size() == 0) {
					level++;
					p = c;
					c = new Stack<>();
				}
			}

		}

	}

	public static void main(String[] args) {
		// TODO Auto-generated method stub
		// 10 true 20 true 40 false false true 50 false false true 30 true 60 false
		// true 74 false true 75 false false true 73 false false
		zigzagtraversal m = new zigzagtraversal();
		BinaryTree tree = m.new BinaryTree();
		tree.display();
		System.out.println();
		tree.zigzag();

	}

}

File: c++/binary tree/zigZagTraversal.cpp
#include <bits/stdc++.h>
using namespace std;

class Node
{
public:
    int data;
    Node *left;
    Node *right;

    Node(int data)
    {
        this->data = data;
        this->left = NULL;
        this->right = NULL;
    }

    ~Node()
    {
        delete left;
        delete right;
    }
};

void display(Node *root)
{
    if (root == NULL)
        return;
    if (root->left)
    {
        cout << root->left->data << " - ";
    }
    else
    {
        cout << ". - ";
    }
    cout << root->data << " - ";
    if (root->right)
    {
        cout << root->right->data << endl;
    }
    else
    {
        cout << "." << endl;
    }
    display(root->left);
    display(root->right);
}

Node *takeInput(vector<int> &arr)
{
    queue<Node *> q;
    Node *root = new Node(arr[0]);
    q.push(root);
    int index = 1;
    while (!q.empty())
    {
        Node *front = q.front();
        q.pop();
        if (arr[index] != -1)
        {
            front->left = new Node(arr[index]);
            q.push(front->left);
        }
        index++;
        if (index == arr.size())
            break;
        if (arr[index] != -1)
        {
            front->right = new Node(arr[index]);
            q.push(front->right);
        }
        index++;
        if (index == arr.size())
            break;
    }
    return root;
}

void zigZagTraversal(Node *root)
{
    stack<Node *> ms;
    stack<Node *> cs;
    int level = 0;
    ms.push(root);
    while (!ms.empty())
    {
        Node *top = ms.top();
        ms.pop();
        cout << top->data << " ";
        if (level % 2 == 0)
        {
            if (top->left)
                cs.push(top->left);
            if (top->right)
                cs.push(top->right);
        }
        else
        {
            if (top->right)
                cs.push(top->right);
            if (top->left)
                cs.push(top->left);
        }
        if (ms.empty())
        {
            cout << endl;
            ms = cs;
            stack<Node *> temp;
            cs = temp;
            level++;
        }
    }
}

int main()
{
    vector<int> arr1 = {50, 25, 75, 12, 37, 62, 87, -1, -1, 30, -1, -1, 70};
    Node *root1 = takeInput(arr1);
    zigZagTraversal(root1);
    delete root1;
    return 0;
}